Color DispersionPlot markers based on target values

lwgray opened the pull request:
This PR creates a Dispersion Plot with color coded classes as requested by PR #549 

If the target values are specified, each word in a document is annotated with the target value associated with the document.  This translates into a colored marker representing the word and the document/class it belongs to

Additionally, when target values are specified, labels appear in a legend with the appropriate markers

![image](https://user-images.githubusercontent.com/1000117/44742576-2ffe9b00-aace-11e8-8f98-c02611219b62.png)


Problem I am facing (Help Needed! 😭 )
-----------------------

I am still faced with the issue of a test passing when text is input as a generator and
no target specified.  Under these conditions, this results in a "ValueError: need at least one array to stack".  This isn't the case when a generator is used with a target specified.  Nor is it the case
when text is input as a list, with or without the target specified. 

Works under these conditions
1.  Input of text as a list and no target is specified
2.  Input of text as a list and target is specified
3.  Input of text as a generator and target is specified

Doesn't work under this condition
1.  Input of text as a generator and no target is specified

rebeccabilbro commented:
Awesome; thanks @lwgray - I'll take a look tomorrow! Can't wait!!

rebeccabilbro commented:
Hey there @lwgray! Looks like your tests are passing now; glad you worked through the bug! Does that mean this is ready for review?

lwgray commented:
It’s ready.

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -58,19 +71,30 @@ class DispersionPlot(TextVisualizer):
+    # NOTE: cannot be np.nan
+    NULL_CLASS = None
+
+    def __init__(self, words, ax=None, colors=None, ignore_case=False,
```
Is it necessary for `words` to be passed in on instantiation? It looks like the user has to pass `text` when they call `fit()`, and that `words` is either the same as `text` or could be derived from it.

lwgray replied:
@rebeccabilbro I am not sure that I understand.   `words` is the list of targeted words that the user wants to be marked throughout the document.  These are the words that appear on the y-axis.

rebeccabilbro replied:
Gotcha! I think I'm getting mixed up because we have both `self.words` and `self.target_words_` here. To make it a bit easier to understand, could we perhaps change `self.words` to `self.target_words` and `self.target_words_` to `self.indexed_words_`, or something like that?

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -41,9 +47,16 @@ class DispersionPlot(TextVisualizer):
+        Specify the colors for each individual class
 
+    colormap : string or matplotlib cmap
+        Sequential colormap for continuous target
```
Continuous targets are associated with regression problems, but I think the goal for this visualizer is to be used on classification problems, right? If the user is going to pass in a colormap here, we may want to suggest a qualitative colormap rather than a sequential one, since that will probably be easier to discern classes.

lwgray replied:
@rebeccabilbro Sorry, The previous wording got carried over from the Tsne visualizer, I will change the wording to reflect your suggestion of using a qualitative colormap...

rebeccabilbro replied:
Awesome, thank you!

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -95,34 +120,91 @@ def fit(self, text):
 
+        # Resolve the labels with the classes
+        labels = self.labels if self.labels is not None else self.classes_
+        if len(labels) != len(self.classes_):
```
Might be a good idea to add a test that make sure this error gets raised in the number of labels doesn't match the number of colors.

lwgray replied:
Maybe I am missing something very obvious.  Won't the number of labels always match the number of color since the number of colors is dictated by the number of labels.  L179-181

rebeccabilbro replied:
Sorry - I meant to type "if the number of labels doesn't match the number of classes". I am referring to you `if` statement on line 172. I was suggesting that you add a test that ensures the error is raised if the number of labels doesn't match the number of classes.

lwgray replied:
I believe I am already catching this mismatch in my tests...  [L115-L129](https://github.com/DistrictDataLabs/yellowbrick/blob/5d6da64aefaa47208fe36db5691ae1200540d3d0/tests/test_text/test_dispersion.py#L115-L129)

rebeccabilbro replied:
How about using a `match` inside your test to ensure that the custom error message you have written for the `YellowbrickValueError` is raised. Something like [this](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/test_classifier/test_classification_report.py#L166-L172) or [this](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/test_draw.py#L33-L34).

lwgray replied:
Got it!

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -95,34 +120,91 @@ def fit(self, text):
+            self.classes_ = np.array([self.NULL_CLASS])
+
         # Create an index (e.g. the y position) for the target words
         self.target_words_ = np.flip(self.words, axis=0)
```
What happens if one or more of the `self.target_words_` isn't in the corpus?

lwgray replied:
@rebeccabilbro Great Catch!  I came across this the other day and forgot about it.  If no target words are in the corpus then an error is raised.  

I now have to decide if I present a blank graph or raise a yellowbrickValueError in this case

rebeccabilbro replied:
As a user, it would be great to get a custom `YellowbrickValueError` in this case! Maybe some thing like "Could not find {} in the corpus."

lwgray replied:
I finally got a fix for this..it is in the latest commit

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -158,14 +245,27 @@ def dispersion(words, corpus, ax=None, color=None,
+        Specify the colors for each individual class
+
+    colormap : string or matplotlib cmap
+        Sequential colormap for continuous target
```
Again, I think this should be qualitative colormap for discrete target

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -58,19 +71,30 @@ class DispersionPlot(TextVisualizer):
-    def _compute_dispersion(self, text):
+    def _compute_dispersion(self, text, y):
         self.boundaries_ = []
         self.offset = 0
```
Does `self.offset` need to be a class attribute? I looked but it doesn't appear to be used outside of the `_compute_dispersion` function.

lwgray replied:
I agree!

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -95,34 +120,91 @@ def fit(self, text):
         # Stack is used to create a 2D array from the generator
-        points = np.stack(self._compute_dispersion(text))
-        self.draw(points)
+        self.points_target = np.stack(self._compute_dispersion(text, y))
```
Does `self.points_target` need to be a class attribute? It doesn't appear to be used outside of `fit()`.

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -95,34 +120,91 @@ def fit(self, text):
-        self.draw(points)
+        self.points_target = np.stack(self._compute_dispersion(text, y))
+
+        self.points = np.stack(zip(self.points_target[:,0].astype(int),
```
It looks like `self.points` could just be `points`, since it isn't used outside of `fit()`.

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -79,13 +103,14 @@ def _compute_dispersion(self, text):
+    def fit(self, text, y=None, **kwargs):
         """
         The fit method is the primary drawing input for the dispersion
         visualization. It requires the corpus as a list of words.
```
It looks like there may have been some changes since the original implementation and might need to update the docstring here; does `DispersionPlot` still expect the input data to be a list of words? It looks like it could either be a list of documents or a generator that would yield a list of documents.

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -79,13 +103,14 @@ def _compute_dispersion(self, text):
         self.boundaries_ = np.array(self.boundaries_, dtype=int)
 
-    def fit(self, text):
+    def fit(self, text, y=None, **kwargs):
```
Would you mind if we changed `text` to `X` to make it a bit more similar to how input features are referenced in scikit-learn and in our other text visualizers (e.g. [FreqDist](http://www.scikit-yb.org/en/develop/_modules/yellowbrick/text/freqdist.html#FrequencyVisualizer.fit) and [TSNE](http://www.scikit-yb.org/en/develop/_modules/yellowbrick/text/tsne.html#TSNEVisualizer.fit)?

rebeccabilbro commented on yellowbrick/text/dispersion.py:
```diff
@@ -95,34 +120,91 @@ def fit(self, text):
         text : list
```
For our other text visualizers, we allow users to pass data to `fit()` as either an ndarray or a dataframe, so we may want to anticipate the case that a user passes in the corpus as a dataframe here,

lwgray replied:
Are you referring to the `y` variable?

rebeccabilbro replied:
I was referring to the `text` (i.e. the corpus data), which is passed to `fit()`.

lwgray replied:
duh! 👎 FAIL

lwgray replied:
How would the dataframe be structured?
rows = documents
columns = words

rebeccabilbro commented:
This is looking great @lwgray - I'm so excited to be able to use the new color-by-class feature! I've added in some inline comments above; please let me know if you have any questions. Are you thinking about updating the `DispersionPlot` docs as part of this PR or of doing that in a subsequent one?

lwgray commented:
@rebeccabilbro I plan on doing a separate PR for the docs as I am quite busy this week.  I don't want to hold up your usage of this Visualizer for your PyData Talks

rebeccabilbro commented:
Ok, no prob! Mind opening up a new issue for the docs (if there isn't already one)?

lwgray commented:
In the last commit,  I address all except 2 of your suggestions.  They require a bit more effort and will be completed ASAP.

To be completed:
-------------------
Test pandas dataframe as input into `fit`
Custom yellowValueError when a word is not found in the corpus

Below are the corrections that were made thus far.
--------------------------------------------------------
1.  Fix colormap docstring entry by changing wording to "Qualitative colormap for discrete targets"
2.  To make it a bit easier to understand, we changed self.words to self.target_words and self.target_words_ to self.indexed_words
3.  Changed name of text variable from 'text' to 'X' to make it a bit more similar to how input features are referenced in scikit-learn and in our other text visualizers (e.g. FreqDist and TSNE)
4. Because `self.offset` is only used in `_compute_dispersion` function it is prudent to change it from class attribute to regular variable `offset`
5.  Modify docstring for the `fit` method to reflect the change in how the method expects input.  It has changed from a list of words to a list/generator of documents that in itself contains a list of words.
6. Because `self.points` and `self.points_target` are only used in `fit` function it is prudent to change it from class attribute to regular variable `points` and `points_target`
7. Create Test that raises if number of labels doesn't match the number of classes

lwgray commented on examples/rebeccabilbro/dispersion_plots.ipynb:
```diff
@@ -0,0 +1,436 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlcAAAFhCAYAAABd8I+pAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzsnXe4VNX1/j8Xpdg7QRMj1lfshhgRC2jsKSpfTSxosILdxFiiWBPyM4k9KnYJGns09hawBqMJRg2Krw3UGBFLNGBBEX5/rD0wDHMv3MuV5vo8z33uzD67n5k566y9z3obJk+eTJIkSZIkSdI6tJnTHUiSJEmSJJmfSOMqSZIkSZKkFUnjKkm ...
```
the variable `chapter_text` within the `fit` method should be `X` instead
